src/backend/distributed/executor/local_executor.c

```diff
@@ -175,6 +178,120 @@ ExecuteLocalTaskList(CitusScanState *scanState, List *taskList)
 }
 
 
+/*
+ * LocalTaskPlannedStmt creates the PlannedStmt for the input task.
+ *
+ * The function behaves differently when the task->queryString is avaliable
+ * or not. See the comments in the function for the details.
+ */
+static Query *
+LocalShardQuery(Query *workerJobQuery, Task *task, ParamListInfo boundParams,
+				int *numParams, Oid **parameterTypes)
+{
+	/*
+	 * For performance reasons, the queryString is not generated for
+	 * local-fast path queries. Instead, we simply update the shard
+	 * references to the actual shard relations.
+	 *
+	 * Otherwise, like multi-shard queries going through the local execution,
+	 * we'd need to parse the queryString back to Query before passing it
+	 * to the planner,
+	 */
+	if (task->query != NULL)
+	{
+		Query *shardQuery = copyObject(workerJobQuery);
+
+		ReplaceShardReferencesWalker((Node *) shardQuery, task);
+
+#if PG_VERSION_NUM < 120000
+		AcquireRewriteLocks(shardQuery, true, false);
```

Comment:
Is it correct that the third argument (`forUpdatePushedDown`) here is always `false`? According to the docs this should always be false for at the top of the recursion. So my main question is:

If you have a subquery in a `FOR UPDATE` query, does the subquery get to this codepath? In that case we don't get strong enough locks I think.

Reply:
What is the reason for having `AcquireRewriteLocks `? When we call postgres standard_planner, wouldn't it call it for us?

Reply:
AcquireRewriteLocks is called (indirectly) by the parser, not the planner. Since now we don't call the parser anymore, we need to acquire the necessary locks manually.

Reply:
To be clear, the locks that we need to acquire are the ones on the shard tables. Since we modify the `relid` fields to those after parsing the locks are never acquired on those tables otherwise.